Add the Motion to the top of the query plan earlier, as a Path.

hlinnaka opened the pull request:
Since PostgreSQL 9.6 and the "upper planner pathification", the planner
tends to do everything with Paths, and only generates the Plan tree at
the end of planning. In the same spirit, add the Motion to the top of
the plan, to bring the result from the QE nodes to the QD, or to take
the result to the right segments if it's INSERT/UPDATE/DELETE, before
converting the Path to Plan tree. The Path tree is generally nicer to
work with, as you have more information available in the Paths. In
particular, we still have the locus information of each Path available.

That's the gist of this PR. It's just one commit in the series, though, there are a bunch of other commits included, which do related cleanup, before and after the main commit. I recommend reviewing the commits one by one.

d commented on src/backend/optimizer/plan/createplan.c:
```diff
@@ -2499,9 +2508,9 @@ create_motion_plan(PlannerInfo *root, CdbMotionPath *path)
+		/* Push the MotionPath's locus down onto subpath. */
 		subpath->locus = path->path.locus;
-		subpath->pathkeys = path->path.pathkeys;
+		//subpath->pathkeys = path->path.pathkeys;
```
commented code: did you mean to delete this?

hlinnaka replied:
Yep, fixed thanks

d commented:
I noticed in a few examples there was a curious plan change (for example in `subselect_gp.out`): A Motion would receive an `initPlan` expression that's identical to its child's. While it is harmless to have duplicate init plans pointing to the same parameter and plan_id, it's kind of wrong (not to mention duplicate initialization during execution). The root cause of that seems to be how `create_motion_plan` and `create_splitupdate_plan` passes the same `root` but with a "subpath" to `create_subplan`. (C.f. The "pathification" change (commit postgres/postgres@68fa28f77146) in 9.6 which introduces `SS_attach_initplans`)

I've been wondering: would those functions be better served by `create_plan_recurse` directly? What is the value of `create_subplan` now?

Jesse and Deep

d commented on src/test/regress/expected/subselect_gp.out:
```diff
@@ -162,37 +162,37 @@ NOTICE:  table "mrs_t1" does not exist, skipping
+                                     QUERY PLAN                                     
+------------------------------------------------------------------------------------
+ Gather Motion 3:1  (slice1; segments: 3)  (cost=3.25..6.85 rows=20 width=4)
+   InitPlan 1 (returns $0)  (slice3)
```
This curious difference is caused by a duplicate initplan

pengzhout commented on src/include/cdb/cdbpathlocus.h:
```diff
@@ -27,6 +27,14 @@ struct PathTarget;
+ * with Replicated locus *must* be executed on all of the segments, whereas
+ * a SegmentGeneral *may* be executed on all of the segments, or just one
+ * of them, as is convenient. Replicated is used as the locus for
+ * UPDATEs/DELETEs/INSERTs to replicated tables; it's important that the
```
Replicated is also applied to the result of broadcast motion mostly.

kainwen replied:
> Replicated is also applied to the result of broadcast motion mostly.

I think it is **only** used for a broadcast motion.

It is a redundant locus type. I think we even can remove it totally.

I have removed it from cdbpath_motion_for_join.

Because broadcast motion cannot be the top plannode in any case.

pengzhout replied:
I don't understand why it's a redundant locus type, even a join path will not have a replicated locus, other paths may still need it to stand for a broadcast motion.

kainwen replied:
> I don't understand why it's a redundant locus type, even a join path will not have a replicated locus, other paths may still need it to stand for a broadcast motion.

Just paste my thought here:

IMO, other locus is the type system of distribute plan. We rely on them to type check the plan.

replicated locus does nothing to help type check a distributed plan.

`Redundant` means we can refactor it out. Other fields might encode the same information for motion type.

hlinnaka replied:
I pushed this comment change from the PR as is. I *think* it's correct, based your comments here. But if I'm wrong, or you can suggest a better wording, please do!

FWIW, I tried to refactor Replicated and SegmentGeneral locus into one, because they seemed redundant to me, too. I hacked on that for a while, but couldn't get it working. That's how I came to the understanding of the difference, that I wrote down in the comment.

pengzhout commented on src/backend/cdb/cdbmutate.c:
```diff
@@ -180,132 +168,6 @@ directDispatchCalculateHash(Plan *plan, GpPolicy *targetPolicy, Oid *hashfuncs)
-
-
 /* -------------------------------------------------------------------------
  * Function apply_motion() and apply_motion_mutator() add motion nodes to a
```
we might need to update comments here.

hlinnaka replied:
Updated, please take a look.

pengzhout commented on src/backend/cdb/cdbmutate.c:
```diff
@@ -2644,8 +2642,6 @@ cdbpathtoplan_create_sri_plan(RangeTblEntry *rte, PlannerInfo *subroot, Path *su
 		/* Build a partitioned flow */
 		resultplan->plan.flow->flotype = FLOW_PARTITIONED;
 		resultplan->plan.flow->locustype = CdbLocusType_Hashed;
-		resultplan->plan.flow->hashExprs = hashExprs;
```
hashExprs do make confuse with the locus. One quick question: with this PR, are all plannings made based on paths now?  Once a plan is created with best_path, it does change, right? otherwise, we have no locus in the plan node to add plan node correctly.

hlinnaka replied:
There is still one place where we rely on the Flow information, and add Motions based on Plan tree rather than Path tree. Namely, in apply_motion() where it processes the SubPlans. If the locus/flow of a SubPlan is not "compatible" with the outer query where the SubPlan is contained, it adds a Broadcast or Gather motion. So we still need the Flow for that, but because that code doesn't care about the distribution keys of the subplan, we don't need the hash keys.

I'd like to refactor that more, so that we would not do that, but I'm not sure how yet. Currently, in make_subplan() we do this (same as in PostgreSQL):

```
	/*
	 * Select best Path and turn it into a Plan.  At least for now, there
	 * seems no reason to postpone doing that.
	 */
	final_rel = fetch_upper_rel(subroot, UPPERREL_FINAL, NULL);
	best_path = get_cheapest_fractional_path(final_rel, tuple_fraction);

	plan = create_plan(subroot, best_path);
```
In GPDB, perhaps we should postpone calling create_plan(), until we know where it appears in the parent plan. That way, we could add the Gather/Broadcast Motion on top of the subplan as a Path, before we call create_plan(). But that seems like a pretty big change, too.

Even if we don't do that, I think we can get rid of the Flows with some other refactoring. For that thing with SubPlans, apply_motion() only needs the flow/locus information at the top of the sub-Plan. We could store that elsewhere, maybe in the 'subroot'. We also use the Flow information to build the slice table in executor startup, so that needs a replacement too. I think a better way to do that would be to build the slice table, or some preliminary representation of the slice table, before the executor. Perhaps in apply_motion().

I'm hacking on those further refactorings, but what we have in this PR is a step in the right direction. We'll see if we can go further than that easily.

pengzhout commented:
On Wed, Nov 13, 2019 at 5:56 PM Heikki Linnakangas <notifications@github.com>
wrote:

> *@hlinnaka* commented on this pull request.
> ------------------------------
>
> In src/backend/cdb/cdbmutate.c
> <https://github.com/greenplum-db/gpdb/pull/9023#discussion_r345663149>:
>
> > @@ -2644,8 +2642,6 @@ cdbpathtoplan_create_sri_plan(RangeTblEntry *rte, PlannerInfo *subroot, Path *su
>  		/* Build a partitioned flow */
>  		resultplan->plan.flow->flotype = FLOW_PARTITIONED;
>  		resultplan->plan.flow->locustype = CdbLocusType_Hashed;
> -		resultplan->plan.flow->hashExprs = hashExprs;
>
> There is still one place where we rely on the Flow information, and add
> Motions based on Plan tree rather than Path tree. Namely, in apply_motion()
> where it processes the SubPlans. If the locus/flow of a SubPlan is not
> "compatible" with the outer query where the SubPlan is contained, it adds a
> Broadcast or Gather motion. So we still need the Flow for that, but because
> that code doesn't care about the distribution keys of the subplan, we don't
> need the hash keys.
>
Thanks for the clarification.

> I'd like to refactor that more, so that we would not do that, but I'm not
> sure how yet. Currently, in make_subplan() we do this (same as in
> PostgreSQL):
>
> 	/*
> 	 * Select best Path and turn it into a Plan.  At least for now, there
> 	 * seems no reason to postpone doing that.
> 	 */
> 	final_rel = fetch_upper_rel(subroot, UPPERREL_FINAL, NULL);
> 	best_path = get_cheapest_fractional_path(final_rel, tuple_fraction);
>
> 	plan = create_plan(subroot, best_path);
>
> In GPDB, perhaps we should postpone calling create_plan(), until we know
> where it appears in the parent plan. That way, we could add the
> Gather/Broadcast Motion on top of the subplan as a Path, before we call
> create_plan(). But that seems like a pretty big change, too.
>
So, is this because the subplan might be made in the pre-process stage and
the outer rel is not even planned yet?

> Even if we don't do that, I think we can get rid of the Flows with some
> other refactoring. For that thing with SubPlans, apply_motion() only needs
> the flow/locus information at the top of the sub-Plan. We could store that
> elsewhere, maybe in the 'subroot'. We also use the Flow information to
> build the slice table in executor startup, so that needs a replacement too.
> I think a better way to do that would be to build the slice table, or some
> preliminary representation of the slice table, before the executor. Perhaps
> in apply_motion().
>
Yes, in this case, plan node doesn't need to maintain the Flow any more, we
can deduce the slice table from motion node and 'info' in subroot.

> I'm hacking on those further refactorings, but what we have in this PR is
> a step in the right direction. We'll see if we can go further than that
> easily.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/greenplum-db/gpdb/pull/9023?email_source=notifications&email_token=ADRWFM5WINQCUTF2YPK6XHLQTPFOFA5CNFSM4JK7SKZKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCLLZSEY#discussion_r345663149>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADRWFM6D4U5QCCINENFHYV3QTPFOFANCNFSM4JK7SKZA>
> .
>

hlinnaka commented:
On 13/11/2019 12:25, Tang Pengzhou wrote:
> So, is this because the subplan might be made in the pre-process stage and
> the outer rel is not even planned yet?

That's right.

- Heikki

hlinnaka commented:
> I noticed in a few examples there was a curious plan change (for example in `subselect_gp.out` : https://github.com/greenplum-db/gpdb/pull/9023/files#r345515063): A Motion would receive an `initPlan` expression that's identical to its child's. While it is harmless to have duplicate init plans pointing to the same parameter and plan_id, it's kind of wrong (not to mention duplicate initialization during execution).

I don't understand. Old plan for that query in subselect_gp test looked like this:
```
create table mrs_t1(x int) distributed by (x);
insert into mrs_t1 select generate_series(1,20);
explain (costs off) select * from mrs_t1 where exists (select x from mrs_t1 where x < -1);
                       QUERY PLAN                       
--------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)
   ->  Result
         One-Time Filter: $0
         InitPlan 1 (returns $0)  (slice3)
           ->  Gather Motion 3:1  (slice2; segments: 3)
                 ->  Seq Scan on mrs_t1 mrs_t1_1
                       Filter: (x < '-1'::integer)
         ->  Seq Scan on mrs_t1
 Optimizer: Postgres query optimizer
(9 rows)
```

New plan:
```
                    QUERY PLAN                    
--------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)
   InitPlan 1 (returns $0)  (slice3)
     ->  Gather Motion 3:1  (slice2; segments: 3)
           ->  Seq Scan on mrs_t1 mrs_t1_1
                 Filter: (x < '-1'::integer)
   ->  Result
         One-Time Filter: $0
         ->  Seq Scan on mrs_t1
 Optimizer: Postgres query optimizer
(9 rows)
```
I don't see any duplicate InitPlans here. The only difference is that the InitPlan was moved from the Result node to the parent Motion node.

hlinnaka commented:
> I've been wondering: would those functions be better served by `create_plan_recurse` directly? What is the value of `create_subplan` now?

Hmm, perhaps. Now that I look at it, I agree it looks pretty sketchy. Especially in create_splitupdate_plan(), it doesn't even save & restore root->curOuterRels/Params like create_motion_plan() does.

d commented:
> I don't see any duplicate InitPlans here. The only difference is that the InitPlan was moved from the Result node to the parent Motion node.

You need to look at the actual plan tree (`SET debug_print_plan TO on`). We have `EXPLAIN` logic that would "de-duplicate" `plan_id`s for `SUBPLAN` expressions -- which is a bit sketchy: while "real" subplan expressions can freely reuse the same `plan_id` (because they need to be re-evaluated usually), I consider it a bug in planner to have duplicate `plan_id` among `InitPlan`'s.

Jesse

hlinnaka commented:
> > I don't see any duplicate InitPlans here. The only difference is that the InitPlan was moved from the Result node to the parent Motion node.
> 
> You need to look at the actual plan tree (`SET debug_print_plan TO on`). We have `EXPLAIN` logic that would "de-duplicate" `plan_id`s for `SUBPLAN` expressions -- which is a bit sketchy: while "real" subplan expressions can freely reuse the same `plan_id` (because they need to be re-evaluated usually), I consider it a bug in planner to have duplicate `plan_id` among `InitPlan`'s.

Ah, gotcha. I agree that's a bug. I'll try to fix it, thanks!

hlinnaka commented:
I pushed the straightforward cleanup commits, to remove now-unused code. I rebased the main commit, to "Add the Motion to the top of the query plan earlier, as a Path" over master, and added a commit to fix the way Init Plans are attached, per the discussion with @d and @soumyadeep2007.

Please take another look!

hlinnaka commented on src/backend/cdb/cdbmutate.c:
```diff
@@ -813,12 +410,32 @@ apply_motion_mutator(Node *node, ApplyMotionState *context)
+		 * we don't recurse into it yet. The subtree is considered part of
+		 * the sending slice, so we need to set up the current context
+		 * accordingly first.
+		 */
```
After hacking some more on this, I realized that the above is not correct. The target list, initPlans, hash expressions, and everything else on the Motion field should be considered be part of the *sender* slice, like the subtree. All of those things are evaluated in the sender.

I'll fix that..

hlinnaka replied:
Fixed. This is ready for review now.

hlinnaka commented:
Pushed, thanks for the reviews, everyone!
